### PR TITLE
[7.0] Revert "Make --incompatible_remote_results_ignore_disk a no-op."

### PR DESCRIPTION
**Pull request by jerrymarino**
We'd like to use Bazel 7.0 but we still rely on this feature for a number of repos.

This issue has a summary of the problems: https://github.com/bazelbuild/bazel/issues/17428

The main changes we'd need are:

1. to filter out several hundreds of GB from the disk cache - doing this many writes caused disk corruption on macOS. #17428 has more info about it

2. When we use this feature with RBE it prevents from uploading artifacts that are tagged as `no-remote-cache`. https://github.com/bazelbuild/bazel-buildfarm/issues/1402 has the issue for buildfarm

*Setting `--incompatible_remote_results_ignore_disk=true` on 6.0, breaks RBE on 6.x.x for us too, because we set many mnemonics which are RBE inputs to `no-remote-cache`. The `no-remote-cache` actions are dependencies of RBE  test actions - but fail to be uploaded to the execution service.

For instance, we have remotely executed `TestRunner a` who depends on `Bundle b`:
```
TestRunner "a" -> Bundle "b"

TestRunner remote
Bundle no-remote-cache
```
`--incompatible_remote_results_ignore_disk=true` causes `Bundle b` to not be uploaded to to the execution service 

This reverts commit c88f374246c0a7b3a87c58553d3b9ad0ef4e9ad4.

### CODE REVIEW

**Comment by jerrymarino** [comment]
cc @lberki @coeuvre  for viz ☝️

We can try to come up with a proposal / fixes for the 2 issues but will probably need more time to prioritize it

**Comment by lberki** [comment]
Do I understand correctly that :

1. There is a data corruption issue with the disk cache and the reason why you are using `--incompatible_remote_results_ignore_disk=false` is that it reduced the likelihood of corruption, but does not eliminate it? [this comment](https://github.com/bazelbuild/bazel/issues/17428#issuecomment-1666627128) seems to imply that this is not the fault of Bazel, but it is still the problem of Bazel...

2. If one sets `--incompatible_remote_results_ignore_disk=true` and there is an RBE action with a `no-remote-cache` dependency, that action will fail on RBE because the `no-remote-cache` dependency does not get uploaded to RBE at all? If so, that sounds like a pretty straightforward bug...

**Comment by jerrymarino** [comment]
In pt1. we eliminated the corruption after a 1.5 weeks of shadow testing, it is totally eliminated, overall it isn’t the sole reason. 

For pt1. / Beyond a flaky file system, these are some additional use cases we wouldn’t want Bazel want to write everything to the disk cache it does locally:
- Too small disks
  - Old laptop,cheap laptop
  - Some cloud devices
- Too slow of disk
  - Slow file system
  - Slow physical device
- Too expensive
  - i/o based price models
- Overhead due to IT spyware/syscall monitoring 

Still Im not sure we can rule out the use case that some systems just cannot cope with this I/o load  in _Bazel_

I think we’d want to better controls for it what gets written, but I’m not sure exactly the details yet hence the desire to sit on it for a while longer.

Yeah on pt2. it’s more of a bug - we’ll just have to get some free cycles time to fix it!

**Comment by lberki** [comment]
Out of curiosity, how did you fix the corruption issue? By upgrading to Mac OS 13.4?

What I'm hearing is that even if the bug in #2 (did you file an issue about it? I haven't found but that doesn't mean there isn't one), controlling the disk cache separately from "local action cache" makes sense.

Based on what I know now, I'm inclined to agree , but I don't know what line of reasoning went into coupling the disk cache with the local action cache and the #15148 doesn't quite explain (it only argues that the disk cache should not be considered a "remote" cache, which is a very good thing). 

The logical conclusion of all this would be a `no-disk-cache` tag; it's Yet Another Knob, but it looks like it's an inevitable conclusion of what you want to do here.

@coeuvre WDYT?

**Comment by jerrymarino** [comment]
> Out of curiosity, how did you fix the corruption issue? By upgrading to Mac OS 13.4?

The TLDR we flipped  `incompatible_remote_results_ignore_disk=false.` where it was originally set to true in 6.0. We didn't update the OS to that version, though it would be seperate to this if it does help with the corruption. 

The failing starlark rule I mentioned was a smoking gun - not the root cause. We had the same problem across stuff running during a build e.g. C++ linking, other python based file creation, protoc ,etc

> The logical conclusion of all this would be a no-disk-cache tag; it's Yet Another Knob, but it looks like it's an inevitable conclusion of what you want to do here.

Yeah I was also thinking something around these lines too, though if we do this I wonder if it can handle - for instance controlling reads/writes to it at a build level as we do for the remote ones:
- `--incompatible_remote_results_ignore_disk=false --remote_upload_results=false # read disk cache`
- `--incompatible_remote_results_ignore_disk=false --remote_upload_results=true # populate disk cache`

**Comment by coeuvre** [comment]
Sorry for the late reply, I was on vacation.

I would rather not reverting the change because disk cache is really **not** a remote cache. Mixing it with remote cache and use `remote-*` tags to control its cache behaviour had caused lots of problems in the past. On top of that, I have some improvements in my mind for the disk cache which are only possible after the "separation".

That's being said, I do want to fully address your issues before 7. Assuming `--incompatible_remote_results_ignore_disk=true`, we have following tags to control the cache behaviours:

- `no-cache`: disable both disk cache and remote cache.
- `no-remote-cache`: disable only remote cache, allow disk cache.

It looks to me that we can replace `no-remote-cache` with `no-cache` to fix point 1. I am not convinced to add yet-another-tag `no-disk-cache` to disable only disk cache.

For point 2, Bazel should always upload inputs to RBE for remote execution no matter whether the action accepts remote/disk cache or not. Otherwise, it is a bug. Do you have a repro?

**Comment by lberki** [comment]
No worries, the important part is that this gets resolved before Bazel 7. If that happens by a means other than merging this pull request, I'm happier because the flag that it resurrects really was a wart.

**Comment by meisterT** [comment]
I assume this PR is obsolete?
